### PR TITLE
ROX-34642: Diff-based endpoint updates in replaceNoLock

### PR DESCRIPTION
**Pull request by vikin91**
## Description

Non-incremental endpoint updates (`Apply(updates, false)`) used to purge all
endpoints for a deployment and reinsert them from scratch on every call, even
when only a single endpoint changed. For deployments behind NodePort services
(200-500 endpoints), this caused thousands of unnecessary allocations per
update cycle, creating GC pressure in Sensor.

This PR replaces the purge-all/reinsert-all cycle with a three-pass diff approach:

1. **Fast path** (`endpointsUnchangedNoLock`) - Added in #20462: O(1)-alloc shortcut when nothing
   changed - the common steady-state case. Uses a stack-allocated hint table
   for target info comparison to avoid heap allocations entirely.
2. **Diff path** (`diffReplaceNoLock`): per-endpoint diff that only touches
   added, removed, or changed endpoints. Mutation cost is O(k × T) where k is
   the number of endpoints that actually changed, not the total M.
   - **Removed endpoints** are moved to history (Pass 1).
   - **New endpoints** are inserted via `insertSingleEndpointNoLock` (Pass 2).
   - **Changed target info** is updated in-place via `clear(etiSet)` + refill,
     avoiding one map delete and one `set.Set` allocation per changed endpoint.
     Old target info is properly recorded in history (Pass 3).
3. **Full insert** (`applySingleNoLock`): only used for brand-new deployments
   that need endpoint-takeover logic.

### Correctness fix: history preservation for changed target info

The old purge-all/reinsert-all path had a latent bug: `addToHistory` recorded
old target info in the history maps, but `deleteFromHistory` inside
`insertSingleEndpointNoLock` immediately removed it, silently discarding the
history entry. The diff path fixes this by calling `addToHistory` without
`deleteFromHistory` for changed target info, so old values are retained in
history until expiry. This scenario (same endpoint IP:port, different container
port or port name) is uncommon in practice but is now handled correctly.

The alternative considered was improving the existing purge/reinsert to allocate
less (e.g., reuse maps), but that still does O(M) history moves per update.
The diff approach avoids the moves entirely for unchanged endpoints.

AI-Assisted: cursor, ~70% generated, user reviewed logic and benchmarks

## User-facing documentation

- [x] [CHANGELOG.md](https://github.com/stackrox/stackrox/blob/master/CHANGELOG.md) is updated **OR** update is not needed
- [x] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

## Testing and quality

- [x] the change is production ready: the change is [GA](https://github.com/stackrox/stackrox/blob/master/PR_GA.md), or otherwise the functionality is gated by a [feature flag](https://github.com/stackrox/stackrox/blob/master/pkg/features/README.md)
- [x] CI results are [inspected](https://docs.google.com/document/d/1d5ga073jkv4CO1kAJqp8MPGpC6E1bwyrCGZ7S5wKg3w/edit?tab=t.0#heading=h.w4ercgtcg0xp)

### Automated testing

- [x] added unit tests
- [ ] added e2e tests
- [x] added regression tests
- [ ] added compatibility tests
- [x] modified existing tests

### How I validated my change

- Unit tests & benchmarks

### Benchstat (10 runs, after rebase onto current master - with history enabled)

```
goos: darwin
goarch: arm64
pkg: github.com/stackrox/rox/sensor/common/clusterentities
cpu: Apple M3 Pro
                                 │ /tmp/bench_rebased_baseline.txt │    /tmp/bench_rebased_branch.txt    │
                                 │             sec/op              │   sec/op     vs base                │
ApplyUnchanged/ep10_tgt2-12                            1.769µ ± 5%   1.675µ ± 1%   -5.29% (p=0.008 n=10)
ApplyUnchanged/ep50_tgt4-12                            6.805µ ± 1%   6.850µ ± 1%        ~ (p=0.363 n=10)
ApplyUnchanged/ep200_tgt4-12                           27.19µ ± 3%   27.22µ ± 1%        ~ (p=0.481 n=10)
ApplyChanged/ep10_tgt2-12                             12.982µ ± 3%   6.530µ ± 2%  -49.71% (p=0.000 n=10)
ApplyChanged/ep50_tgt4-12                              70.15µ ± 1%   36.77µ ± 1%  -47.59% (p=0.000 n=10)
ApplyChanged/ep200_tgt4-12                             285.1µ ± 0%   150.7µ ± 1%  -47.14% (p=0.000 n=10)
ApplyPartialChange/ep50_tgt2-12                        64.18µ ± 1%   11.31µ ± 2%  -82.38% (p=0.000 n=10)
ApplyPartialChange/ep200_tgt2-12                      267.43µ ± 0%   45.43µ ± 2%  -83.01% (p=0.000 n=10)
ApplyPartialChange/ep200_tgt4-12                      294.92µ ± 1%   50.84µ ± 2%  -82.76% (p=0.000 n=10)
ApplyPartialChange/ep500_tgt2-12                       714.4µ ± 2%   119.8µ ± 1%  -83.23% (p=0.000 n=10)
geomean                                                56.07µ        22.64µ       -59.63%

                                 │ /tmp/bench_rebased_baseline.txt │     /tmp/bench_rebased_branch.txt      │
                                 │              B/op               │     B/op      vs base                  │
ApplyUnchanged/ep10_tgt2-12                             672.0 ± 0%     672.0 ± 0%        ~ (p=1.000 n=10) ¹
ApplyUnchanged/ep50_tgt4-12                             672.0 ± 0%     672.0 ± 0%        ~ (p=1.000 n=10) ¹
ApplyUnchanged/ep200_tgt4-12                            672.0 ± 0%     672.0 ± 0%        ~ (p=1.000 n=10) ¹
ApplyChanged/ep10_tgt2-12                            15.271Ki ± 0%   2.553Ki ± 0%  -83.28% (p=0.000 n=10)
ApplyChanged/ep50_tgt4-12                            72.687Ki ± 0%   9.022Ki ± 0%  -87.59% (p=0.000 n=10)
ApplyChanged/ep200_tgt4-12                           288.35Ki ± 0%   34.66Ki ± 0%  -87.98% (p=0.000 n=10)
ApplyPartialChange/ep50_tgt2-12                       74180.0 ± 0%     742.0 ± 0%  -99.00% (p=0.000 n=10)
ApplyPartialChange/ep200_tgt2-12                     294269.0 ± 0%     742.0 ± 0%  -99.75% (p=0.000 n=10)
ApplyPartialChange/ep200_tgt4-12                     295075.0 ± 0%     746.0 ± 0%  -99.75% (p=0.000 n=10)
ApplyPartialChange/ep500_tgt2-12                     821684.0 ± 0%     742.0 ± 0%  -99.91% (p=0.000 n=10)
geomean                                               29.10Ki        1.512Ki       -94.80%
¹ all samples are equal

                                 │ /tmp/bench_rebased_baseline.txt │     /tmp/bench_rebased_branch.txt     │
                                 │            allocs/op            │  allocs/op   vs base                  │
ApplyUnchanged/ep10_tgt2-12                             4.000 ± 0%    4.000 ± 0%        ~ (p=1.000 n=10) ¹
ApplyUnchanged/ep50_tgt4-12                             4.000 ± 0%    4.000 ± 0%        ~ (p=1.000 n=10) ¹
ApplyUnchanged/ep200_tgt4-12                            4.000 ± 0%    4.000 ± 0%        ~ (p=1.000 n=10) ¹
ApplyChanged/ep10_tgt2-12                              128.00 ± 0%    44.00 ± 0%  -65.62% (p=0.000 n=10)
ApplyChanged/ep50_tgt4-12                               692.0 ± 0%    286.0 ± 0%  -58.67% (p=0.000 n=10)
ApplyChanged/ep200_tgt4-12                             2.721k ± 0%   1.113k ± 0%  -59.10% (p=0.000 n=10)
ApplyPartialChange/ep50_tgt2-12                       567.000 ± 0%    8.000 ± 0%  -98.59% (p=0.000 n=10)
ApplyPartialChange/ep200_tgt2-12                     2221.000 ± 0%    8.000 ± 0%  -99.64% (p=0.000 n=10)
ApplyPartialChange/ep200_tgt4-12                      2621.00 ± 0%    10.00 ± 0%  -99.62% (p=0.000 n=10)
ApplyPartialChange/ep500_tgt2-12                     5525.000 ± 0%    8.000 ± 0%  -99.86% (p=0.000 n=10)
geomean                                                 221.3         18.46       -91.66%
¹ all samples are equal
```

#### Benchmark explanation

**ApplyUnchanged** is flat — the fast-path (endpointsUnchangedNoLock) is already
on master and this PR does not change that code path.

**ApplyChanged** (all endpoints have different target info) improves because the
diff path updates each endpoint's target info set in-place via clear+refill
instead of the old purge-all/reinsert-all cycle. On master, every Apply call
deletes all endpoints from the current maps, allocates a fresh set.Set for
each one, re-adds them, and performs a redundant add+delete on the history
maps. The diff path skips all of that: the endpoint never leaves the current
maps, no new set.Set is allocated (the existing map buckets are reused), and
history is recorded once without being immediately discarded. This eliminates
~6 map operations and 1 heap allocation per endpoint per update cycle.

**ApplyPartialChange** (1 out of N endpoints changed) improves the most because
the diff path skips unchanged endpoints entirely. On master, even a single
changed endpoint triggers a full purge and reinsert of all N endpoints. The
diff path only touches the one endpoint that actually changed — the other
N-1 endpoints are compared in O(T) each (via targetInfoUnchangedNoLock) and
left in place with zero mutations. The cost drops from O(N × T) mutations
to O(N × T) read-only comparisons plus O(1 × T) mutations.

### CODE REVIEW

**Comment by vikin91** [comment]
This change is part of the following stack:

- #20462
    - #20628 ◀

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

**Comment by vikin91** [comment]
@coderabbitai review

**Comment by vikin91** [comment]
@coderabbitai review

File: sensor/common/clusterentities/store_endpoints.go [hunk 1]
```diff
@@ -137,9 +142,140 @@ func (e *endpointsStore) replaceNoLock(updates map[string]*EntityData) {
...
+	dSet.Add(ep)
+
+	if _, ok := e.endpointMap[ep]; !ok {
+		e.endpointMap[ep] = make(map[string]set.Set[EndpointTargetInfo], 1)
+	}
+	etiSet := make(set.Set[EndpointTargetInfo], len(targetInfos))
+	for _, ti := range targetInfos {
+		etiSet.Add(ti)
+	}
+	e.endpointMap[ep][deploymentID] = etiSet
```
Comment (janisz) [inline sensor/common/clusterentities/store_endpoints.go]: Can we get rid of make and use map initialization?
Also we can move this lines below just where they are about to be used

File: sensor/common/clusterentities/store_endpoints.go [hunk 2]
```diff
@@ -137,9 +142,140 @@ func (e *endpointsStore) replaceNoLock(updates map[string]*EntityData) {
...
+}
+
+// insertSingleEndpointNoLock inserts one endpoint's target info for an existing
+// deployment. Unlike applySingleNoLock, it does not perform deployment-takeover
+// checks (those only apply when a deployment is first seen).
+func (e *endpointsStore) insertSingleEndpointNoLock(deploymentID string, ep net.NumericEndpoint, targetInfos []EndpointTargetInfo) {
+	dSet := e.reverseEndpointMap[deploymentID]
+	if dSet == nil {
+		dSet = make(set.Set[net.NumericEndpoint], 1)
+		e.reverseEndpointMap[deploymentID] = dSet
```
Comment (janisz) [inline sensor/common/clusterentities/store_endpoints.go:218]:
```suggestion
		dSetd = set.Set[net.NumericEndpoint]{dSet: struct{}{}}
```

File: sensor/common/clusterentities/store_endpoints.go [hunk 3]
```diff
@@ -137,9 +142,140 @@ func (e *endpointsStore) replaceNoLock(updates map[string]*EntityData) {
...
+	for ti := range currentTargetInfos {
+		idx := int(hints[ti.ContainerPort%hintBuckets]) - 1
+		if idx >= 0 && idx < n && newTargetInfos[idx] == ti {
+			continue
+		}
+		if !slices.Contains(newTargetInfos, ti) {
+			return false
+		}
+	}
+	return true
```
Comment (janisz) [inline sensor/common/clusterentities/store_endpoints.go]: It's  super edge case do we need it? Maybe the basic aproach will be OK. Could you add a benchmark to compare where hints performs better than naive appraoch.

File: sensor/common/clusterentities/store_endpoints.go [hunk 4]
```diff
@@ -137,9 +142,140 @@ func (e *endpointsStore) replaceNoLock(updates map[string]*EntityData) {
...
+// endpointsUnchangedNoLock. Extracting a shared helper made local
+// BenchmarkApply{Unchanged,Changed} runs about 3-7% slower and
+// BenchmarkEndpointsUnchangedNoLock about 3% slower, with no B/op or
+// allocs/op improvement.
+func (e *endpointsStore) targetInfoUnchangedNoLock(deploymentID string, ep net.NumericEndpoint, newTargetInfos []EndpointTargetInfo) bool {
+	currentTargetInfos, found := e.endpointMap[ep][deploymentID]
+	if !found {
+		return len(newTargetInfos) == 0
+	}
+	if len(currentTargetInfos) != len(newTargetInfos) {
```
Comment (janisz) [inline sensor/common/clusterentities/store_endpoints.go:254]: let's add a test for currentTargetInfos, newTargetInfos empty in different configurations.

File: sensor/common/clusterentities/store_endpoints.go [hunk 5]
```diff
@@ -137,9 +142,140 @@ func (e *endpointsStore) replaceNoLock(updates map[string]*EntityData) {
...
+// slice of equal length.
+//
+// The comparison logic intentionally duplicates the inner hot path from
+// endpointsUnchangedNoLock. Extracting a shared helper made local
+// BenchmarkApply{Unchanged,Changed} runs about 3-7% slower and
+// BenchmarkEndpointsUnchangedNoLock about 3% slower, with no B/op or
+// allocs/op improvement.
+func (e *endpointsStore) targetInfoUnchangedNoLock(deploymentID string, ep net.NumericEndpoint, newTargetInfos []EndpointTargetInfo) bool {
+	currentTargetInfos, found := e.endpointMap[ep][deploymentID]
+	if !found {
```
Comment (janisz) [inline sensor/common/clusterentities/store_endpoints.go]: We can replace the rest of code with Equal:

https://cs.opensource.google/go/go/+/refs/tags/go1.26.3:src/slices/slices.go;l=20

**Review by janisz — APPROVED** [review]

File: sensor/common/clusterentities/store_endpoints.go (hunk 5 above)
Comment (vikin91) [reply sensor/common/clusterentities/store_endpoints.go]: slices.Equal requires sorting. I went with the simple Contains. PTAL

**Review by janisz — APPROVED** [review]

File: sensor/common/clusterentities/store_endpoints.go (hunk 1 above)
Comment (vikin91) [reply sensor/common/clusterentities/store_endpoints.go]: Addressed by using in place initialization

File: sensor/common/clusterentities/store_endpoints.go (hunk 3 above)
Comment (vikin91) [reply sensor/common/clusterentities/store_endpoints.go]: Dropped that as we will never have more target infos per endpoint than 4. And if we do then we can pay the performance price.
